File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/trigger/Trigger.java
package com.raizlabs.android.dbflow.sql.trigger;

import com.raizlabs.android.dbflow.sql.Query;
import com.raizlabs.android.dbflow.sql.QueryBuilder;
import com.raizlabs.android.dbflow.sql.language.property.IProperty;
import com.raizlabs.android.dbflow.structure.Model;

/**
 * Description: Describes an easy way to create a SQLite TRIGGER
 */
public class Trigger implements Query {

    /**
     * Specifies that we should do this TRIGGER before some event
     */
    public static final String BEFORE = "BEFORE";

    /**
     * Specifies that we should do this TRIGGER after some event
     */
    public static final String AFTER = "AFTER";

    /**
     * Specifies that we should do this TRIGGER instead of the specified events
     */
    public static final String INSTEAD_OF = "INSTEAD OF";

    /**
     * The name in the DB
     */
    final String triggerName;

    /**
     * If it's {@link #BEFORE}, {@link #AFTER}, or {@link #INSTEAD_OF}
     */
    String beforeOrAfter;

    boolean temporary;

    /**
     * @param triggerName The name of the trigger to use.
     * @return A new trigger.
     */
    public static Trigger create(String triggerName) {
        return new Trigger(triggerName);
    }

    /**
     * Creates a trigger with the specified trigger name. You need to complete
     * the trigger using
     *
     * @param triggerName What we should call this trigger
     */
    private Trigger(String triggerName) {
        this.triggerName = triggerName;
    }

    /**
     * Sets the trigger as temporary.
     */
    public Trigger temporary() {
        this.temporary = true;
        return this;
    }

    /**
     * Specifies AFTER eventName
     */
    public Trigger after() {
        beforeOrAfter = AFTER;
        return this;
    }

    /**
     * Specifies BEFORE eventName
     */
    public Trigger before() {
        beforeOrAfter = BEFORE;
        return this;
    }

    /**
     * Specifies INSTEAD OF eventName
     */
    public Trigger insteadOf() {
        beforeOrAfter = INSTEAD_OF;
        return this;
    }

    /**
     * Starts a DELETE ON command
     *
     * @param onTable The table ON
     */
    public <TModel extends Model> TriggerMethod<TModel> delete(Class<TModel> onTable) {
        return new TriggerMethod<>(this, TriggerMethod.DELETE, onTable);
    }

    /**
     * Starts a INSERT ON command
     *
     * @param onTable The table ON
     */
    public <TModel extends Model> TriggerMethod<TModel> insert(Class<TModel> onTable) {
        return new TriggerMethod<>(this, TriggerMethod.INSERT, onTable);
    }

    /**
     * Starts an UPDATE ON command
     *
     * @param onTable    The table ON
     * @param properties if empty, will not execute an OF command. If you specify columns,
     *                   the UPDATE OF column1, column2,... will be used.
     */
    public <TModel extends Model> TriggerMethod<TModel> update(Class<TModel> onTable, IProperty... properties) {
        return new TriggerMethod<>(this, TriggerMethod.UPDATE, onTable, properties);
    }

    /**
     * @return The name of this TRIGGER
     */
    public String getName() {
        return triggerName;
    }

    @Override
    public String getQuery() {
        QueryBuilder queryBuilder = new QueryBuilder("CREATE ");
        if (temporary) {
            queryBuilder.append("TEMP ");
        }
        queryBuilder.append("TRIGGER IF NOT EXISTS ")
            .appendQuotedIfNeeded(triggerName).appendSpace()
            .appendOptional(" " + beforeOrAfter + " ");

        return queryBuilder.getQuery();
    }
}

File: lib/src/main/kotlin/com/dbflow5/query/Trigger.kt
package com.dbflow5.query

import com.dbflow5.adapter2.DBRepresentable
import com.dbflow5.appendQuotedIfNeeded
import com.dbflow5.database.DatabaseWrapper
import com.dbflow5.query.operations.AnyOperator
import com.dbflow5.query.operations.Property
import com.dbflow5.sql.Query

internal sealed class TriggerQualifier(val value: String) {
    object Before : TriggerQualifier("BEFORE")
    object After : TriggerQualifier("AFTER")
    object InsteadOf : TriggerQualifier("INSTEAD OF")

    companion object {
        val All = listOf(Before, After, InsteadOf)
    }
}

internal sealed class TriggerMethod(val value: String) {
    object Insert : TriggerMethod("INSERT")
    data class Update<Table : Any>(val properties: List<Property<*, Table>>) :
        TriggerMethod("UPDATE")

    object Delete : TriggerMethod("DELETE")
    companion object {
        val All = listOf(Insert, Update<Any>(listOf()), Delete)
    }
}

interface IsTriggerQualified<Table : Any> {
    /**
     * Specifies that we should do this TRIGGER before some event
     */
    fun before(): TriggerQualified<Table>

    /**
     * Specifies that we should do this TRIGGER after some event
     */
    fun after(): TriggerQualified<Table>

    /**
     * Specifies that we should do this TRIGGER instead of the specified events
     */
    fun insteadOf(): TriggerQualified<Table>
}

interface TriggerMethodEnabled<Table : Any> {
    fun insertOn(): TriggerOn<Table>

    /**
     * If [ofProperties] specified, this UPDATE becomes UPDATE OF p0,p1,...
     */
    fun updateOn(vararg ofProperties: Property<*, Table>): TriggerOn<Table>
    fun deleteOn(): TriggerOn<Table>
}

interface TriggerForEachRowEnabled<Table : Any> {
    fun forEachRow(): TriggerForEachRow<Table>
}

interface TriggerWhenEnabled<Table : Any> {
    infix fun whenever(operator: AnyOperator): TriggerWhen<Table>
}

interface TriggerLogicEnabled<Table : Any> {
    infix fun begin(query: Query): TriggerLogic<Table>
}

interface Trigger<Table : Any> : Query,
    HasAdapter<Table, DBRepresentable<Table>>

interface TriggerStart<Table : Any> : Trigger<Table>,
    IsTriggerQualified<Table>,
    TriggerMethodEnabled<Table>

interface TriggerQualified<Table : Any> : Trigger<Table>,
    TriggerMethodEnabled<Table>

interface TriggerOn<Table : Any> : Trigger<Table>,
    TriggerForEachRowEnabled<Table>,
    TriggerWhenEnabled<Table>,
    TriggerLogicEnabled<Table>

interface TriggerForEachRow<Table : Any> : Trigger<Table>,
    TriggerWhenEnabled<Table>,
    TriggerLogicEnabled<Table>

interface TriggerWhen<Table : Any> : Trigger<Table>,
    TriggerLogicEnabled<Table>

interface TriggerLogic<Table : Any> : Trigger<Table>,
    ExecutableQuery<Unit> {
    infix fun and(query: Query): TriggerLogic<Table>
}

/**
 * Starts a TRIGGER statement.
 */
fun <Table : Any> DBRepresentable<Table>.createTrigger(
    name: String,
    temporary: Boolean = false,
    ifNotExists: Boolean = true,
): TriggerStart<Table> {
    return TriggerImpl(
        name = name,
        temporary = temporary,
        ifNotExists = ifNotExists,
        adapter = this,
    )
}

internal data class TriggerImpl<Table : Any>(
    override val adapter: DBRepresentable<Table>,
    private val name: String,
    private val temporary: Boolean,
    private val ifNotExists: Boolean,
    private val qualifier: TriggerQualifier? = null,
    private val method: TriggerMethod? = null,
    private val forEachRow: Boolean = false,
    private val whenOperator: AnyOperator? = null,
    private val triggerLogicQuery: List<Query> = emptyList(),
) : TriggerStart<Table>,
    TriggerQualified<Table>,
    TriggerOn<Table>,
    TriggerForEachRow<Table>,
    TriggerWhen<Table>,
    TriggerLogic<Table> {
    override val query: String by lazy {
        buildString {
            append("CREATE ")
            if (temporary) append("TEMP ")
            append("TRIGGER ")
            if (ifNotExists) append("IF NOT EXISTS ")
            appendQuotedIfNeeded(name).append(" ")
            if (qualifier != null) append("${qualifier.value} ")
            if (method != null) append("${method.value} ")
            if (method is TriggerMethod.Update<*> &&
                method.properties.isNotEmpty()
            ) {
                append("OF ${method.properties.joinToString { it.query }} ")
            }
            append("ON ${adapter.name} ")
            if (forEachRow) append("FOR EACH ROW ")
            whenOperator?.let { append("WHEN ${it.query} ") }
            if (triggerLogicQuery.isNotEmpty()) {
                append(
                    "\nBEGIN\n${
                        triggerLogicQuery.joinToString(separator = ";\n") { it.query }
                    };\nEND"
                )
            }
        }
    }

    override fun insertOn(): TriggerOn<Table> =
        copy(
            method = TriggerMethod.Insert,
        )

    override fun updateOn(vararg ofProperties: Property<*, Table>): TriggerOn<Table> =
        copy(
            method = TriggerMethod.Update(ofProperties.toList())
        )

    override fun deleteOn(): TriggerOn<Table> =
        copy(
            method = TriggerMethod.Delete,
        )

    override fun before(): TriggerQualified<Table> =
        copy(
            qualifier = TriggerQualifier.Before,
        )

    override fun after(): TriggerQualified<Table> =
        copy(
            qualifier = TriggerQualifier.After,
        )

    override fun insteadOf(): TriggerQualified<Table> =
        copy(
            qualifier = TriggerQualifier.InsteadOf,
        )

    override fun forEachRow(): TriggerForEachRow<Table> =
        copy(
            forEachRow = true,
        )

    override fun whenever(operator: AnyOperator): TriggerWhen<Table> =
        copy(
            whenOperator = operator,
        )

    override fun begin(query: Query): TriggerLogic<Table> =
        copy(
            triggerLogicQuery = listOf(query),
        )

    override fun and(query: Query): TriggerLogic<Table> =
        copy(
            triggerLogicQuery = triggerLogicQuery.toMutableList().apply { add(query) }
        )

    override suspend fun execute(db: DatabaseWrapper) {
        db.execSQL(query)
    }
}